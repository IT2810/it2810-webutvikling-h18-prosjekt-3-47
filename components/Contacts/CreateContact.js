/**
 * CreateContact displays a contact form.
 * When pushing the save (after some input checks) a contact with the given information is sent back to the
 * ContactsScreen, via the callback function, for storing. After the information is used in ContactScreen, a callback
 * is sent back to CreateContact to notify that the input (via state) can be safely cleared.
 */
import React from 'react'
import { StyleSheet, Text, View, TextInput, Button, Alert } from 'react-native';
import c from '../../constants/Colors';

export default class CreateContact extends React.Component {
    constructor (props) {
        super(props);
        this.state = {
            name: null,
            phoneNumber: null,
            address: null
        };

        this.clearInputs = this.clearInputs.bind(this);
    }
    /**
     * Sends the state (a new contact) back to the receiveNewContact-method in ContactScreen, if all fields are set
     * Then clears the textInputs and dismisses the popup, to make ready for new contact creations
     */
    sendContact() {
        //console.log(this.state);
        if (this.state.name && this.state.phoneNumber && this.state.address) {
            if (this.state.name.trim().length > 0 && this.state.phoneNumber.trim().length > 0 && this.state.address.trim().length > 0) {

                this.props.callback(this.state, this.clearInputs);
                this.props.popupDialog.dismiss();
            }
            else {
                // Some field contain only whitespace
                Alert.alert('Warning','Input length must be more than 0')
            }
        }
        else {
            // Not all fields are filled
            Alert.alert('Warning', 'All fields must be filled');
        }
    }

    clearInputs() {
        this.setState({
            address: null,
            phoneNumber: null,
            name: null
        });
    }

    render() {
        return (
            <View style={styles.container}>
                <Text style={styles.text}>Name:</Text>
                <TextInput
                    style={styles.input}
                    value={this.state.name}
                    onChangeText={(text) => this.setState({name: text})}
                    placeholder="Fill in name here"
                />
                <Text style={styles.text}>Phone Number:</Text>
                <TextInput
                    keyboardType={'numeric'}
                    style={styles.input}
                    value={this.state.phoneNumber}
                    onChangeText={(text) => this.setState({phoneNumber: text})}
                    placeholder="Fill in phone number here"
                />
                <Text style={styles.text}>Address:</Text>
                <TextInput
                    style={styles.input}
                    value={this.state.address}
                    onChangeText={(text) => this.setState({address: text})}
                    placeholder="Fill in address here"
                />
                <View style={styles.button}>
                    <Button
                    title={'  Save  '}
                    onPress={() => {
                        this.sendContact();
                    }}
                    />
                </View>
            </View>
        );
    }
}
const styles = StyleSheet.create({
    container: {
        backgroundColor: c.appMainBackground,
        justifyContent: 'center',
    },
    input: {
        height: 40,
        width: '94%',
        marginLeft: '3%',
    },
    text: {
        marginLeft: '3%',
        fontSize: 16,
        marginTop: '2%',
    },
    button: {
        alignItems: 'center',
        marginTop: '2%',
    }
});
