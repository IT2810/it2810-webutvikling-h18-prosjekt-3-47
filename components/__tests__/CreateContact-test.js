import 'react-native';
import React from 'react';
import CreateContact from '../Contacts/CreateContact';
import renderer from 'react-test-renderer';
import ShallowRenderer from 'react-test-renderer/shallow';

jest.mock("react-native-popup-dialog");

let shallowRenderer = new ShallowRenderer();

let contacts = [
    {
        name: "Navn Navnesen 1",
        phoneNumber: 12345678,
        address: "Adresse 1"
    },
    {
        name: "Navn Navnesen 2",
        phoneNumber: 87654321,
        address: "Adresse 2"
    }
];

describe('Rendering', () => {
    it('renders correctly', () => {
        const tree = renderer.create(<CreateContact callback={null} popupDialog={null}/>).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('renders shallow correctly', () => {
        shallowRenderer.render(<CreateContact callback={null} popupDialog={null}/>);
        const result = shallowRenderer.getRenderOutput();

        expect(result).toMatchSnapshot();
    });
});

describe('Logical', () => {
    it('should have all information set to null on first render', () => {

        let createContact = renderer.create(<CreateContact callback={null} popupDialog={null}/>);
        let createContactInstance = createContact.getInstance();

        expect(createContactInstance.state.name).toBeNull();
        expect(createContactInstance.state.phoneNumber).toBeNull();
        expect(createContactInstance.state.address).toBeNull();
    });

    it('should clear state with the clearInputs() method', () => {

        let createContact = renderer.create(<CreateContact callback={null} popupDialog={null}/>);
        let createContactInstance = createContact.getInstance();

        let contact = contacts[0];

        createContactInstance.setState({name: contact.name});

        expect(createContactInstance.state.name).toEqual(contact.name);
        expect(createContactInstance.state.phoneNumber).toBeNull();
        expect(createContactInstance.state.address).toBeNull();

        createContactInstance.clearInputs();

        expect(createContactInstance.state.name).toBeNull();
        expect(createContactInstance.state.phoneNumber).toBeNull();
        expect(createContactInstance.state.address).toBeNull();
    });
});

