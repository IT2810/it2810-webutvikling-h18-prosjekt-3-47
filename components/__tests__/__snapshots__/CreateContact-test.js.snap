// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "justifyContent": "center",
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 16,
        "marginLeft": "3%",
        "marginTop": "2%",
      }
    }
  >
    Name:
  </Text>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="Fill in name here"
    style={
      Object {
        "height": 40,
        "marginLeft": "3%",
        "width": "94%",
      }
    }
    value={null}
  />
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 16,
        "marginLeft": "3%",
        "marginTop": "2%",
      }
    }
  >
    Phone Number:
  </Text>
  <TextInput
    allowFontScaling={true}
    keyboardType="numeric"
    onChangeText={[Function]}
    placeholder="Fill in phone number here"
    style={
      Object {
        "height": 40,
        "marginLeft": "3%",
        "width": "94%",
      }
    }
    value={null}
  />
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 16,
        "marginLeft": "3%",
        "marginTop": "2%",
      }
    }
  >
    Address:
  </Text>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="Fill in address here"
    style={
      Object {
        "height": 40,
        "marginLeft": "3%",
        "width": "94%",
      }
    }
    value={null}
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "marginTop": "2%",
      }
    }
  >
    <View
      accessibilityComponentType="button"
      accessibilityTraits={
        Array [
          "button",
        ]
      }
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
            ]
          }
        >
            Save  
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`Rendering renders shallow correctly 1`] = `
<Component
  style={
    Object {
      "backgroundColor": "#fff",
      "justifyContent": "center",
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 16,
        "marginLeft": "3%",
        "marginTop": "2%",
      }
    }
  >
    Name:
  </Text>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="Fill in name here"
    style={
      Object {
        "height": 40,
        "marginLeft": "3%",
        "width": "94%",
      }
    }
    value={null}
  />
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 16,
        "marginLeft": "3%",
        "marginTop": "2%",
      }
    }
  >
    Phone Number:
  </Text>
  <TextInput
    allowFontScaling={true}
    keyboardType="numeric"
    onChangeText={[Function]}
    placeholder="Fill in phone number here"
    style={
      Object {
        "height": 40,
        "marginLeft": "3%",
        "width": "94%",
      }
    }
    value={null}
  />
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 16,
        "marginLeft": "3%",
        "marginTop": "2%",
      }
    }
  >
    Address:
  </Text>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="Fill in address here"
    style={
      Object {
        "height": 40,
        "marginLeft": "3%",
        "width": "94%",
      }
    }
    value={null}
  />
  <Component
    style={
      Object {
        "alignItems": "center",
        "marginTop": "2%",
      }
    }
  >
    <Button
      onPress={[Function]}
      title="  Save  "
    />
  </Component>
</Component>
`;
